src/hexdump.rs

```diff
@@ -0,0 +1,241 @@
+use crate::document::BytesFormat;
+use crate::error::Error;
+use regex::RegexBuilder;
+
+const HEX: &[u8; 16] = b"0123456789abcdef";
+
+// Emit bytes as a hex string (e.g. "cafef00d0badc0de").
+fn hexstr(data: &[u8]) -> String {
+    let mut s = String::with_capacity(2 * data.len());
+    for byte in data {
+        s.push(HEX[(byte >> 4) as usize] as char);
+        s.push(HEX[(byte & 0x0F) as usize] as char);
+    }
+    s
+}
+
+// Emit bytes as a hexdump in the style of `hexdump -vC`.
+fn hexdump(data: &[u8]) -> String {
+    let mut s = String::with_capacity(5 * data.len());
+    for (i, chunk) in data.chunks(16).enumerate() {
+        if i > 0 {
+            s.push('\n');
+        }
+        s.push_str(&format!("{:08x}", i * 16));
+        let mut buf = [b'.'; 16];
+        let mut space = 51;
+        for (j, &byte) in chunk.iter().enumerate() {
+            if j % 8 == 0 {
+                s.push(' ');
+                space -= 1;
+            }
+            s.push(' ');
+            s.push(HEX[(byte >> 4) as usize] as char);
+            s.push(HEX[(byte & 0x0F) as usize] as char);
+            space -= 3;
+            buf[j] = if byte >= 0x20 && byte < 0x7f {
+                byte
+            } else {
+                b'.'
+            };
+        }
+        // Utf8Error is impossible here because all of the codepoints
+        // inside `buf` are ASCII.
+        let chars = std::str::from_utf8(&buf[..chunk.len()]).unwrap();
+        s.push_str(&format!("{0:>1$} |{2}|", " ", space, chars));
+    }
+    s
+}
```

Comment:
`hexdump` uses `String::with_capacity`, which is usually used to avoid allocations, but the `5 * data.len()` value is not well justified. For example, if `data.len() == 1`, then only `5` bytes will be allocated, but an entire line will be pushed into `s`.

`write!(s, "...", ...).unwrap()` instead of `s.push_str(&format!("...", ...))` will avoid allocating a new `String` and copying it into `s`.

Also, this can be simplified by separating "print the hexadecimal digits" and "print the characters or dots" into two loops.

Perhaps this implementation is clearer?

```rust
fn hexdump(data: &[u8]) -> String {
    // 00000000  54 68 65 20 71 75 69 63  6b 20 62 72 6f 77 6e 20  |The quick brown |\n
    // AAAAAAAAABBBBBBBBBBBBBBBBBBBBBBBBCBBBBBBBBBBBBBBBBBBBBBBBBDDDEEEEEEEEEEEEEEEEFFF
    let mut out = String::new();
    for (row, chunk) in data.chunks(16).enumerate() {
        // A
        write!(out, "{:08x} ", 16 * row).unwrap();
        for col in 0..16 {
            // C
            if col == 8 {
                out.push(' ');
            }
            // B
            match chunk.get(col) {
                Some(byte) => write!(out, " {:02x}", byte).unwrap(),
                None => out.push_str("   "),
            }
        }
        // D
        out.push_str("  |");
        for &byte in chunk {
            // E
            match byte {
                0x20..=0x7e => out.push(byte as char),
                _ => out.push('.'),
            }
        }
        // F
        out.push_str("|\n");
    }
    out.pop(); // Remove trailing newline if not empty
    out
}
```

Reply:
The value of 5 is well justified:
```
for N in `seq 0 4096`; do dd if=/dev/zero bs=1 count=$N 2>/dev/null | hexdump -vC | wc | awk "{print $N, \$3-9, \"ratio =\", (\$3-9)/$N}"; done
```
`hexdump -vC` will always emit 9 extra bytes after the hexdump (the total length dumped followed by a newline, hence the `-9` in the calculation above).
The ratio of output to input converges to 79/16 (approximately 5).  For small input lengths, you may require a few calls to the allocator, but they'll be cheap.  For moderate input lengths (~ >250), you'll seldom require more than one call to the allocator and for large lengths (~ >1024) you'll never require another call.

The calculation can be made more precise by changing from `len*5` to `(len+15)*79/16`, rounding the input length up to the next 16 bytes so that the allocation is always enough to complete a line of output.

I'd forgotten that `String` is really `Vec<u8>` underneath and that I can use `write!` on a string.

Reply:
I still find the implementation less clear than my version, but perhaps that is a personal preference thing.